server/src/main/java/org/candlepin/async/JobManager.java

```diff
@@ -14,45 +14,140 @@
 import org.candlepin.common.filter.LoggingFilter;
+import org.candlepin.config.ConfigProperties;
+import org.candlepin.controller.ModeChangeListener;
+import org.candlepin.controller.ModeManager;
 import org.candlepin.guice.CandlepinRequestScope;
 import org.candlepin.guice.PrincipalProvider;
 import org.candlepin.model.AsyncJobStatus;
 import org.candlepin.model.AsyncJobStatus.JobState;
 import org.candlepin.model.AsyncJobStatusCurator;
+import org.candlepin.model.CandlepinModeChange;
+import org.candlepin.model.CandlepinModeChange.Mode;
 import org.candlepin.util.Util;
 import org.jboss.resteasy.spi.ResteasyProviderFactory;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+
+import com.google.inject.Injector;
+import com.google.inject.persist.UnitOfWork;
+
+import org.quartz.CronTrigger;
+import org.quartz.CronScheduleBuilder;
+import org.quartz.Job;
+import org.quartz.JobDetail;
+import org.quartz.JobKey;
+import org.quartz.ScheduleBuilder;
+import org.quartz.Scheduler;
+import org.quartz.SchedulerFactory;
+import org.quartz.SchedulerException;
+import org.quartz.Trigger;
+import org.quartz.TriggerBuilder;
+import org.quartz.TriggerKey;
+import org.quartz.impl.matchers.GroupMatcher;
+import org.quartz.spi.JobFactory;
+import org.quartz.spi.TriggerFiredBundle;
+
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 import org.slf4j.MDC;
 
-import javax.inject.Inject;
-import javax.inject.Singleton;
 import java.util.Collections;
 import java.util.Date;
 import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
 import java.util.Map;
 import java.util.Objects;
+import java.util.Set;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+
 
 
 /**
  * The JobManager manages the queueing, execution and general bookkeeping on jobs
  */
 @Singleton
-public class JobManager {
+public class JobManager implements ModeChangeListener {
+    private static final Logger log = LoggerFactory.getLogger(JobManager.class);
 
-    private static Logger log = LoggerFactory.getLogger(JobManager.class);
+    private static final String QRTZ_GROUP_CONFIG = "cp_async_config";
+    private static final String QRTZ_GROUP_MANUAL = "cp_async_manual";
+
+    // Temporary constant for the principal key; used to store the principal in the job data
+    public static final String PRINCIPAL_KEY = "principal_key";
 
     /** Stores our mapping of job keys to job classes */
-    private static Map<String, Class<? extends AsyncJob>> jobs = new HashMap<>();
+    private static final Map<String, Class<? extends AsyncJob>> JOB_KEY_MAP = new HashMap<>();
+
+    /**
+     * Enum representing known manager states, and valid state transitions
+     */
+    public static enum ManagerState {
+        // Impl note: We have to use strings here since we can't reference enums that haven't yet
+        // been defined. This is slightly less efficient than I'd like, but whatever.
```

Comment:
Yeah, we definitely don't want (easily) mutable enums here, as that's a lot of extra work to save not much time in the actual check. Fortunately, due to the state names chosen, the vast majority of comparisons will only check the first character of the string, maybe second in some cases; so it's not as inefficient as it would initially seem. A handful of instructions worse than the raw enum comparison.